arbor/threading/threading.hpp

```diff
@@ -97,10 +97,27 @@ class task_system {
 
 class task_group {
 private:
+    struct exception_struct {
+        std::atomic<bool> bail{false};
+        std::exception_ptr exception_ptr_;
+        std::mutex exception_mutex_;
+
+        bool have_exception() {
+            return bail.load(std::memory_order_relaxed);
+        }
+
+        void set_exception(std::exception_ptr ex) {
+            bail.store(true, std::memory_order_relaxed);
+            lock ex_lock{exception_mutex_};
+            exception_ptr_ = std::move(ex);
+        }
+    };
+
```

Comment:
I would propose the following:

* add a `get` method that will return the `exception_ptr` and reset `bail`  to false.
* make the destructor explicitly `std::rethrow_exception` on the `exception_pointer` if `bail` is still set.

This would mean that if users ignore the returned exception state, and it has been set, the exception is thrown for them.

Reply:
Is this in relation to your earlier comment: "An alternative, would be to make the destructor of exception_struct rethrow the exception if it has not explicitly been acquired or rethrown.
That way, if we ignore the return value of task_group::wait(), any exception will be rethrown when it is destructed." ? 
If yes, do you mean we should call the get() method inside wait() ? Because that wouldn't solve the problem. If the user ignores the returned exception state, the destructor still won't throw the exception. 
Or are you saying that we would expose the get() method to the user, and wait() would just return void? Or are you saying something completely different?